### PR TITLE
bitswap_test: make racy test less racy

### LINKED ISSUE #4108: Random test failure in bitswap
#### Version information:

> go-ipfs version: 0.4.11-dev-06c567d17
Repo version: 5
System version: amd64/linux
Golang version: go1.8.3

#### Type:

Bug

#### Severity:

Medium

#### Description:

```
ERROR   bitswap: nil cid in GetBlock get.go:18

--- FAIL: TestBasicBitswap (0.03s)

	bitswap_test.go:319: Test a one node trying to get one block from another

	bitswap_test.go:338: should have no items in other peers wantlist

FAIL

FAIL	github.com/ipfs/go-ipfs/exchange/bitswap	10.387s
```

### PR DESCRIPTION
**Pull request by Stebalien**
fixes #4108

### CODE REVIEW

**Comment by Stebalien** [comment]
(in an effort to make CI reliable...)

**Comment by Stebalien** [comment]
As usual, the failures are unrelated... (99 failures of tests on CI, 99 failures of tests, take one down, pass it around, 9001 failures of tests on CI).

**Comment by Stebalien** [comment]
> We should also object tests that depend on flat timing in future.

YES! Also, a rule that all CI failures must be investigated/reported, even if they look spurious. Always having to re-run CI a few times to get a pass is time consuming (and dangerous).

**Comment by Stebalien** [comment]
Speaking of which... the travis test failure was #4062.

**Comment by whyrusleeping** [comment]
we should add a more generic `waitFor(ctx, check)` method so we can do things like this more frequently.

**Comment by Stebalien** [comment]
Need to:
1. Copy the modifications made to the vendored testutils to the actual library.
2. Unvendor testutils.
3. Use WaitFor from the unvendored testutils.

**Comment by whyrusleeping** [comment]
@Stebalien update here?

**Comment by Stebalien** [comment]
No change. I'll do this now... (it's just really tedious).

**Comment by Stebalien** [comment]
Blocked on https://github.com/ipfs/go-ipfs/pull/4091 apparently (I need to update go-libp2p which now pulls in go-libp2p-circuit).

**Comment by vyzo** [comment]
Note that #4091 is still in circuit-relay-1.1.2, it will need to update to circuit-relay-1.1.3 when libp2p is updated.

File: core/core_test.go [hunk 1]
```diff
@@ -38,7 +38,7 @@ func TestInitialization(t *testing.T) {
 	for i, c := range good {
 		r := &repo.Mock{
 			C: *c,
-			D: testutil.ThreadSafeCloserMapDatastore(),
+			D: ds2.ThreadSafeCloserMapDatastore(),
```
Comment (Stebalien) [inline core/core_test.go]: Note: I moved this function from the vendored testutil library to datastore2.

**Comment by Kubuxu** [comment]
Thank you for this.

**Comment by Kubuxu** [comment]
Thank you a lot for this fix.

**Pull request merged**